### ISSUE TITLE
goto ""; behaves like exit (only if argument is constant empty string)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12004]
Migrated from [rt.perl.org#111794](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=111794) (status was 'resolved')

Searchable as RT111794$

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

#### Created by ambrus@math.bme.hu




The statement \`goto ""' behaves like \`exit'\.  I think it should instead die
with an informative message just like how a goto to a non\-existing label
name does\.

For example\, the following does not output anything&#8203;:

 perl \-we 'goto ""; warn "hello";'


More details\.  This seems to happen only if the empty string argument is a
constant \(after optimizations\)\.  The statement behaves like exit in that
it does not throw an exception eval could catch\, but destructors and END
blocks are still ran\.  This error is quite old\, it seems to exist since at
least perl 5\.8\.something\, and is still there in perl 5\.14\.2 and the latest
blead\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.15.8:

Configured by ambrus at Fri Mar 16 13:07:07 CET 2012.

Summary of my perl5 (revision 5 version 15 subversion 8) configuration:
  Commit id: 48100be180e1ed747313758d3ed70dd9c8879f3e
  Platform:
    osname=linux, osvers=2.6.37, archname=x86_64-linux
    uname='linux king 2.6.37 #6 smp sun mar 13 20:15:05 cet 2011
x86_64 gnulinux '
    config_args='-Dprefix=~/local/perlblead -Dusedevel -d'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector
-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.1', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/../lib /usr/lib/../lib /lib /usr/lib
/lib64 /usr/lib64 /usr/local/lib64
    libs=-lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.11.3.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.11.3'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib
-fstack-protector'

Locally applied patches:



@INC for perl 5.15.8:
    /home/ambrus/local/perlblead/lib/site_perl/5.15.8/x86_64-linux
    /home/ambrus/local/perlblead/lib/site_perl/5.15.8
    /home/ambrus/local/perlblead/lib/5.15.8/x86_64-linux
    /home/ambrus/local/perlblead/lib/5.15.8
    /home/ambrus/local/perlblead/lib/site_perl
    .


Environment for perl 5.15.8:
    HOME=/home/ambrus
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=hu_HU
    LD_LIBRARY_PATH=/home/ambrus/local/lib/
    LOGDIR (unset)
    PATH=/home/ambrus/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Fri\, Mar 16\, 2012 at 07&#8203;:43&#8203;:10AM \-0700\, Zsban Ambrus wrote&#8203;:

> For example\, the following does not output anything&#8203;:
> 
>  perl \-we 'goto ""; warn "hello";'
> 
> 
> More details\.  This seems to happen only if the empty string argument is a
> constant \(after optimizations\)\.  The statement behaves like exit in that
> it does not throw an exception eval could catch\, but destructors and END
> blocks are still ran\.  This error is quite old\, it seems to exist since at
> least perl 5\.8\.something\, and is still there in perl 5\.14\.2 and the latest
> blead\.

Seems to exist in perl 5\.000\.

I don't think Perl 4 allowed goto with a string\, so it seems to be a Perl 5
bug\. \(And as it's in 5\.000\, not a regression\)

Nicholas Clark

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Mar 16 09&#8203;:43&#8203;:49 2012\, nicholas wrote&#8203;:
> On Fri\, Mar 16\, 2012 at 07&#8203;:43&#8203;:10AM \-0700\, Zsban Ambrus wrote&#8203;:
> 
> > For example\, the following does not output anything&#8203;:
> > 
> >  perl \-we 'goto ""; warn "hello";'
> > 
> > 
> > More details\.  This seems to happen only if the empty string
argument is a
> > constant \(after optimizations\)\.  The statement behaves like exit in that
> > it does not throw an exception eval could catch\, but destructors and END
> > blocks are still ran\.  This error is quite old\, it seems to exist
since at
> > least perl 5\.8\.something\, and is still there in perl 5\.14\.2 and the
latest
> > blead\.
> 
> Seems to exist in perl 5\.000\.
> 
> I don't think Perl 4 allowed goto with a string\, so it seems to be a
Perl 5
> bug\. \(And as it's in 5\.000\, not a regression\)
> 
> Nicholas Clark
> 

There is a fix for this on the sprout/misc\-post\-5\.16 branch\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed by c8f85248c6\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
